giraffe/src/components/Geo.tsx

```diff
@@ -0,0 +1,237 @@
+import Control from 'react-leaflet-control'
+import 'leaflet/dist/leaflet.css'
+
+// Components
+import CircleMarkerLayer from './geo/CircleMarkerLayer'
+import HeatmapLayer from './geo/HeatmapLayer'
+import PointMapLayer from './geo/PointMapLayer'
+import TrackMapLayer from './geo/TrackMapLayer'
+
+// Utils
+import {preprocessData} from './geo/processing/tableProcessing'
+
+// Types
+import {
+  GeoCircleViewLayer,
+  GeoHeatMapViewLayer,
+  GeoLayerConfig,
+  GeoPointMapViewLayer,
+  GeoTrackMapViewLayer,
+  GeoViewLayer,
+} from '../types/geo'
+import {Config, Table} from '../types'
+
+// Constants
+const ZOOM_FRACTION = 8
+
+interface Props extends Partial<GeoLayerConfig> {
+  width: number
+  height: number
+  table: Table
+  stylingConfig: Partial<Config>
+}
+
+const getRowLimit = (layers: GeoViewLayer[]) => {
+  return Math.min.apply(
+    null,
+    layers.map(l => {
+      switch (l.type) {
+        case 'circleMap':
+          return 5000
+        case 'heatmap':
+          return 100000
+        case 'pointMap':
+          return 2000
+        case 'trackMap':
+          return 2000
+      }
+    })
+  )
+}
+
+const getMinZoom = (width: number): number => {
+  // Math.max(Math.log2(width/256)),Math.log2(height/256))
+  // while the formula above would be technically correct, problem is that
+  // web map projection is square (as opposed to regular book based world maps).
+  // The polar areas are extremely distorted and people don't
+  // want to look at those - they usually want to see all the continents and
+  // expect to see a rectangle, similar to book based maps.
+  return Math.ceil(Math.log2(width / 256) * ZOOM_FRACTION) / ZOOM_FRACTION
+}
+
+const Geo: FunctionComponent<Props> = props => {
+  const {width, height} = props
+  if (width === 0 || height === 0) return null
+  const {lat, lon, zoom, mapStyle, stylingConfig, allowPanAndZoom} = props
+  const {layers, tileServerConfiguration} = props
+  const {tileServerUrl, bingKey} = tileServerConfiguration
+  const mapRef = React.createRef()
+
+  useEffect(() => {
+    ;(mapRef.current as any).leafletElement._onResize()
+  }, [width, height])
+
+  const {table, detectCoordinateFields} = props
+  const [preprocessedTable, setPreprocessedTable] = useState(
+    table
+      ? preprocessData(
+          table,
+          getRowLimit(props.layers),
+          detectCoordinateFields,
+          preprocessedTable => {
+            setPreprocessedTable(preprocessedTable)
+          }
+        )
+      : null
+  )
+
+  useEffect(() => {
+    const newTable = preprocessData(
+      props.table,
+      getRowLimit(props.layers),
+      props.detectCoordinateFields,
+      preprocessedTable => {
+        setPreprocessedTable(preprocessedTable)
+      }
+    )
+    if (newTable) setPreprocessedTable(newTable)
+  }, [table, detectCoordinateFields])
+
+  const onViewportChange = (viewport: {center?: number[]; zoom?: number}) => {
+    const {onViewportChange} = props
+    if (onViewportChange)
```

Comment:
Please always use braces for if statements. We will be updating the eslint rules to reflect this.